scio-bigquery/src/main/scala/com/spotify/scio/bigquery/nio/BigQueryIO.scala

```diff
@@ -0,0 +1,286 @@
+}
+
+/**
+   * Get an SCollection for a BigQuery TableRow JSON file.
+   * @group input
+   */
+final case class TableRowJsonFile(path: String) extends ScioIO[TableRow] {
+  type ReadP = Unit
+  type WriteP = Nothing // TODO
+
+  def id: String = path
+
+  def read(sc: ScioContext, params: ReadP): SCollection[TableRow] =
+    sc.requireNotClosed {
+      sc.wrap(sc.applyInternal(gio.TextIO.read().from(path))).setName(path)
+        .map(e => util.ScioUtil.jsonFactory.fromString(e, classOf[TableRow]))
+    }
+
+  def tap(read: ReadP): Tap[TableRow] = ???
+  def write(data: SCollection[TableRow], params: WriteP): Future[Tap[TableRow]] = ???
+}
+
+object Typed {
+
+  /**
+   * Get a typed SCollection for a BigQuery table.
+   *
+   * Note that `T` must be annotated with
+   * [[com.spotify.scio.bigquery.types.BigQueryType.fromTable BigQueryType.fromTable]]
+   *
+   * The source (table) specified in the annotation will be used
+   *
+   * {{{
+   * @BigQueryType.fromTable("publicdata:samples.gsod")
+   * class Row
+   * }}}
+   *
+   */
+  def apply[T <: HasAnnotation with HasTable : ClassTag : TypeTag]: ScioIO.Aux[T, Unit, Nothing] = {
+    val bqt = BigQueryType[T]
+    val _table: String = bqt.table.get
+    table(_table)
+  }
+
+  /**
+   * Get a typed SCollection for a BigQuery SELECT query
+   *
+   * Note that `T` must be annotated with
+   * [[com.spotify.scio.bigquery.types.BigQueryType.fromQuery BigQueryType.fromQuery]]
+   *
+   * Both [[https://cloud.google.com/bigquery/docs/reference/legacy-sql Legacy SQL]] and
+   * [[https://cloud.google.com/bigquery/docs/reference/standard-sql/ Standard SQL]] dialects are
+   * supported. By default the query dialect will be automatically detected. To override this
+   * behavior, start the query string with `#legacysql` or `#standardsql`.
+   */
+  def apply[T <: HasAnnotation with HasQuery : ClassTag : TypeTag](implicit dummy: DummyImplicit): ScioIO.ReadOnly[T, Unit] = {
+    val bqt = BigQueryType[T]
+    val _query = bqt.query.get
+    query(_query)
+  }
+
+  /**
+   * Get a typed SCollection for a BigQuery SELECT query
+   *
+   * Both [[https://cloud.google.com/bigquery/docs/reference/legacy-sql Legacy SQL]] and
+   * [[https://cloud.google.com/bigquery/docs/reference/standard-sql/ Standard SQL]] dialects are
+   * supported. By default the query dialect will be automatically detected. To override this
+   * behavior, start the query string with `#legacysql` or `#standardsql`.
+   */
+  def query[T <: HasAnnotation : ClassTag : TypeTag](query: String): ScioIO.ReadOnly[T, Unit] =
+    new ScioIO[T] {
+      type ReadP = Unit
+      type WriteP = Nothing // ReadOnly
+
+      def id: String = query
+      def read(sc: ScioContext, params: ReadP): SCollection[T] = {
+        @inline def typedRead(sc: ScioContext) = Reads.avroBigQueryRead[T](sc)
+        Reads.bqReadQuery(sc)(typedRead(sc), query)
+      }
+
+      def write(data: SCollection[T], params: WriteP): Future[Tap[T]] = ???
```

Comment:
Some IOs like SQL queries are actually read-only. `ScioIO` more or less assume that you can always read and write.
I'm using a little trick here to make `write` impossible to call (you can't pass it an instance of `Nothing`), but splitting `ScioIO` into `Read` and `Write` would probably be a better option. I'd like to limit the scope of this already pretty big PR. I think splitting the traits should be done in another PR.